kernel/http_client/src/lib.rs

```diff
@@ -89,237 +81,261 @@ impl HttpResponse {
     }
 }
 
-
-/// A convenience struct that packages together a connected TCP socket
-/// with other elements that are necessary to transceive packets. 
-pub struct ConnectedTcpSocket<'i, 's, 'sockset_a, 'sockset_b, 'sockset_c> {
-    iface:   &'i NetworkInterfaceRef, 
-    sockets: &'s mut SocketSet<'sockset_a, 'sockset_b, 'sockset_c>,
-    handle:  SocketHandle,
+pub struct HttpClient<'a> {
+    interface: &'a Arc<NetworkInterface>,
+    socket: Socket<tcp::Socket<'static>>,
 }
-impl<'i, 's, 'sockset_a, 'sockset_b, 'sockset_c> ConnectedTcpSocket<'i, 's, 'sockset_a, 'sockset_b, 'sockset_c> {
-    /// Create a new `ConnectedTcpSocket` with the given necessary items:
-    /// # Arguments
-    /// * `iface`: a reference to the `NetworkInterface` that the given TCP socket was created on and uses for transceiving packets. 
-    /// * `sockets`: the set of sockets that includes the given TCP socket (usually just a set with just that one socket).
-    /// * `tcp_socket_handle`: the handle of the TCP socket, which must be in the given `sockets` set and be already connected to the remote endpoint.
-    /// 
-    /// Returns an `Err` result if the TCP socket isn't connected to the remote endpoint.
-    /// 
-    pub fn new(
-        iface: &'i NetworkInterfaceRef, 
-        sockets: &'s mut SocketSet<'sockset_a, 'sockset_b, 'sockset_c>,
-        tcp_socket_handle: SocketHandle, 
-    ) -> Result<ConnectedTcpSocket<'i, 's, 'sockset_a, 'sockset_b, 'sockset_c>, &'static str> {
-        // ensure the socket actually connected to the remote endpoint (i.e., it should be able to send/recv)
-        {
-            let socket = sockets.get::<TcpSocket>(tcp_socket_handle);
-            let connected = socket.may_send() && socket.may_recv();
-            if !connected {
-                return Err("http_client: the given TCP socket wasn't connected to the remote endpoint");
-            }
-        }
 
-        Ok(ConnectedTcpSocket {
-            iface,
-            sockets,
-            handle: tcp_socket_handle,
-        })
+impl<'a> HttpClient<'a> {
+    // TODO: Use per-request destination rather than per-client.
+    /// Creates a new HTTP client connected to the given remote endpoint.
+    pub fn new(
+        interface: &'a Arc<NetworkInterface>,
+        local_port: u16,
+        remote_endpoint: IpEndpoint,
+    ) -> Result<Self, &'static str> {
+        let rx_buffer = tcp::SocketBuffer::new(vec![0; 256]);
+        let tx_buffer = tcp::SocketBuffer::new(vec![0; 256]);
```

Comment:
does the newer smoltcp API require passing in growable storage like a Vec, or can we use a fixed-size slice/array?
If it does require or support a growable collection type, is there a reason we choose a 256-byte zeroed vector rather than an empty one?

Reply:
The storage type can either be a [mutable slice or a vec](https://docs.rs/managed/0.8.0/managed/enum.ManagedSlice.html).

Using a slice wouldn't work because then the lifetime of the socket (and by extension `HttpClient`) would be tied to the lifetime of the array created in the function, and so we wouldn't be able to return it.

The fact that `Vec` is growable isn't important, AFAICT it's only used as a fixed-size owned storage of bytes, so we need to initialise it with a specific size.

Reply:
ah yes, ManagedSlice. Okay thanks, we'll keep it as is then.